src/main/java/hu/oe/nik/szfmv/automatedcar/bus/userinput/KeyboardUserInput.java

```diff
@@ -0,0 +1,216 @@
+    }
+
+    @Override
+    public void unsubscribeSteeringEvents(ISteeringEventHandler handler) {
+        this.steeringEventHandlers.remove(handler);
+    }
+
+    @Override
+    public void unsubscribeShiftingEvents(IShiftingEventHandler handler) {
+        this.shiftingEventHandlers.remove(handler);
+    }
+
+    @Override
+    public void unsubscribeIndicationEvents(IIndicationEventHandler handler) {
+        this.indicationEventHandlers.remove(handler);
+    }
+
+    @Override
+    public void keyTyped(KeyEvent e) {}
+
+    @Override
+    public void keyPressed(KeyEvent e) {
+
+        if(!ensureKey(e.getKeyCode()))
+            handleFilteredKeyEvent(e);
+    }
+
+    @Override
+    public void keyReleased(KeyEvent e) {
+
+        if(e.getKeyCode() == GAS)
+            this.onPedalRelease(PedalType.Gas);
+        else if(e.getKeyCode() == BRAKE)
+            this.onPedalRelease(PedalType.Brake);
+        else if(e.getKeyCode() == STEER_LEFT)
+            this.onSteeringRelease();
+        else if(e.getKeyCode() == STEER_RIGHT)
+            this.onSteeringRelease();
+
+        removeKey(e.getKeyCode());
+    }
+
+
+    private void onPedalPush(PedalType p){
+
+        if(p == PedalType.Gas)
+            gasPedalEventHandlers.forEach(IPedalEventHandler::onPedalPush);
+        else if(p == PedalType.Brake)
+            brakePedalEventHandlers.forEach(IPedalEventHandler::onPedalPush);
+        else
+            throw new UnsupportedOperationException("There isn't event handling support for this pedal type: "+p.name());
+    }
+    private void onPedalRelease(PedalType p){
+
+        if(p == PedalType.Gas)
+            gasPedalEventHandlers.forEach(IPedalEventHandler::onPedalRelease);
+        else if(p== PedalType.Brake)
+            brakePedalEventHandlers.forEach(IPedalEventHandler::onPedalRelease);
+        else
+            throw new UnsupportedOperationException("There isn't event handling support for this pedal type: "+p.name());
+    }
+    private void onSteering(Direction d){ this.steeringEventHandlers.forEach(handler -> handler.onSteering(d)); }
+    private void onSteeringRelease(){ this.steeringEventHandlers.forEach(ISteeringEventHandler::onSteeringReleased); }
+    private void onShifting(Gear g){ this.shiftingEventHandlers.forEach(handler -> handler.onShifting(g));
+    }
+    private void onIndication(Direction d) { this.indicationEventHandlers.forEach(handler -> handler.onIndication(d)); }
+
+    private void argumentCheck(Object o){
+        if(o==null)
+            throw new IllegalArgumentException("The given argument cannot be null!");
+    }
+    private boolean ensureKey(int code){
+        if(!this.holdKeys.contains(code)) {
+            this.holdKeys.add(code);
+            return false;
+        }
+        return true;
+    }
+    private void removeKey(int code){
+        this.holdKeys.remove(this.holdKeys.indexOf(code));
+    }
+    private void handleFilteredKeyEvent(KeyEvent e){
```

Comment:
Azon túl, hogy ez egy igen hosszú metódus, amelynek a komplexitása is elég nagy felvetődik bennem a kérdés, hogy csak akkor akkor lehet jobbra tekerni a kormányt ha nem nyomom a gázt?

Ha a _gáz_ nincs benyomva és a a _fék_ sincs benyomva, akkor fog a balra kanyarodás érvényre lépni. Jól értem?

Ha jól értem a kódot, akkor az else-ekre semmi szükség nincsen. Két egymás alatti if simán lehet a kódban. Nem kell mindenhova else-eket írni.

Reply:
**Billentyűkezelés kérdése:** Az említett "handleFilteredKeyEvent" metódus paramétere az éppen lenyomott billentyű, tehát nem az összes nyomva tartott billentyűt tartalmazza. Ez a metódus eldönti, hogy milyen esemény (gázadás/fékezés/stb.) tartozik az adott billentyűhöz. Ez a függvény tehát nem akadályozza azt, hogy egyszerre több gomb hatása érvényesüljön.

**Tesztelés**: A tesztek hiányosságát elismerem, igyekszem ezen változtatni, illetve köszi, hogy belinkelted a lefedettség mérőt, ez tényleg elkerülte a figyelmem.

**Kódformázás**: Valóban csak a fontosabb dolgokra koncentráltam, feedback-et pedig csupán a CodeFactor nyújtott, de azt is már csak a pull request-nél. Kipróbálom majd a CheckStyle plugin-t, abban viszont tudnál/tudnátok segíteni, hogy hol férek hozzá a szükséges ruleset-hez importálható formában?

**Technikai kérdés**: Mi a "javítás menete"? Ilyenre gondolok, hogy pl. indítok egy új ágat annál a commit-nál ahonnan lett indítva ez a PR és onnan nyitok új pull request-et? Probléma ugye az hogy a csapatunk ágára már kerültek fel új dolgok, így tisztán ezekre a fájlokra onnan nem lehet majd új PR-et nyitni.

Reply:
@dezsi96 

- billentyűkezelés: oké értem.
- kódformázás: A ruleset checkstyle.xml néven a projekt gyökerében található, beállítani [így kell](https://szfmv2018-osz.github.io/handout/fejlesztoi_eszkozok/#checkstyle-plugin)
- javítás: A GH úgy kezeli a PR-eket, hogy ha frissül az a branch amiről indítottad, akkor a PR is automatikusan frissül, újrafuttatva az ellenőrzéseket is. Tehát semmi egyeben nem kell tenned, mint a forrás branchre betolni a javítást.